### PR TITLE
Added example for preference

### PR DESCRIPTION
**Pull request by ajithkumar-maragathavel**
## Purpose of this pull request

This PR will provide an example of how to use the preference node for overriding methods from the core class.

## Affected DevDocs pages

-  https://devdocs.magento.com/guides/v2.3/extension-dev-guide/build/di-xml-file.html

### CODE REVIEW

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md [hunk 1]
```diff
@@ -267,6 +267,67 @@ This mapping is in `app/etc/di.xml`, so the object manager injects the `Magento\
 
 This mapping is in `app/code/Magento/Backend/etc/adminhtml/di.xml`, so the object manager injects the `Magento\Backend\Model\Url` implementation class wherever there is a request for the `Magento\Core\Model\UrlInterface` in the [admin](https://glossary.magento.com/admin) area.
 
+### Override method using Preference
+
+If you want to override a method from core file, you can utilize preference to acheive it.
```
Comment (rogyar) [inline src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: Please, clarify the method scopes that may be overridden 

```suggestion
If you want to override a public or protected method from core class, you can utilize `preference` to achieve it.
```

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md [hunk 2]
```diff
@@ -267,6 +267,67 @@ This mapping is in `app/etc/di.xml`, so the object manager injects the `Magento\
...
+    protected function isVisible(Order $order)
+    {
+        # Write your custom logic here.
+        return !in_array(
+            $order->getStatus(),
+            $this->_orderConfig->getInvisibleOnFrontStatuses()
+        );
+    }
+}
+```
```
Comment (rogyar) [inline src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: Please, add the following callout here:

```
Overriding entire methods is not recommended approach and should be avoided whenever it's possible. This approach brings potential conflicts to the system and reduces the system upgradability. Use other extensibility options like event observers and plugins instead. 
```

It's a good thing to have information about the overriding but we should not advocate it somehow. 

Thank you!

Comment (ajithkumar-maragathavel) [reply src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: @rogyar I will update the docs. That is a great suggestion. Thanks.

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md (hunk 1 above)
Comment (ajithkumar-maragathavel) [reply src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: @rogyar Updated the contents as per your suggestion.

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md [hunk 3]
```diff
@@ -267,6 +267,70 @@ This mapping is in `app/etc/di.xml`, so the object manager injects the `Magento\
 
 This mapping is in `app/code/Magento/Backend/etc/adminhtml/di.xml`, so the object manager injects the `Magento\Backend\Model\Url` implementation class wherever there is a request for the `Magento\Core\Model\UrlInterface` in the [admin](https://glossary.magento.com/admin) area.
 
+### Override method using Preference
+
+If you want to override a public or protected method from core class, you can utilize `preference` to achieve it.
```
Comment (rogyar) [inline src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]:
```suggestion
If you want to override a public or protected method from a core class, you can utilize the `preference` node from `di.xml` to achieve it.
```

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md [hunk 4]
```diff
@@ -267,6 +267,70 @@ This mapping is in `app/etc/di.xml`, so the object manager injects the `Magento\
...
+Here is an example to overide a method from core file.
```
Comment (rogyar) [inline src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]:
```suggestion
Here is an example of overriding a method from the core file.
```

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md [hunk 5]
```diff
@@ -267,6 +267,70 @@ This mapping is in `app/etc/di.xml`, so the object manager injects the `Magento\
...
+
+```xml
+<!-- app/code/<VendorName/<ModuleName>/etc/di.xml -->
```
Comment (rogyar) [inline src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: I would suggest using `OrangeCompany/OverrideExample` as a module/vendor name. See [other](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/routing.html#before-and-after-parameters) documentation for more examples

Comment (rogyar) [reply src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: It applies to all your examples

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md [hunk 6]
```diff
@@ -267,6 +267,70 @@ This mapping is in `app/etc/di.xml`, so the object manager injects the `Magento\
...
+        # Write your custom logic here.
+        return !in_array(
+            $order->getStatus(),
+            $this->_orderConfig->getInvisibleOnFrontStatuses()
+        );
+    }
+}
+```
+
+`Note:`
```
Comment (rogyar) [inline src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: Please use `{:.bs-callout-info}` directive for the callouts. You can see an example of a callout in the [following](https://devdocs.magento.com/guides/v2.3/get-started/authentication/gs-authentication-session.html) topic (the blue section with the "i" icon)

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md (hunk 3 above)
Comment (ajithkumar-maragathavel) [reply src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: @rogyar Updated the content as per your suggestion. Thanks for the review.

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md (hunk 4 above)
Comment (ajithkumar-maragathavel) [reply src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: @rogyar Updated the content and commit the changes. Thanks for the suggestion.

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md (hunk 5 above)
Comment (ajithkumar-maragathavel) [reply src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: @rogyar Thanks for providing an example link. Hereafter will follow the same. Content Updated.

File: src/guides/v2.3/extension-dev-guide/build/di-xml-file.md (hunk 6 above)
Comment (ajithkumar-maragathavel) [reply src/guides/v2.3/extension-dev-guide/build/di-xml-file.md]: @rogyar Fixed this one too.

**Comment by dobooth** [comment]
Hi @ajithkumar-maragathavel and @rogyar. Thanks for the work on this. We are a bit confused here why we are telling them how to do this and then saying that it is not the recommended way. If it is not recommended, then perhaps it should not be in Devdocs. Is there a good reason?

**Comment by ajithkumar-maragathavel** [comment]
> Hi @ajithkumar-maragathavel and @rogyar. Thanks for the work on this. We are a bit confused here why we are telling them how to do this and then saying that it is not the recommended way. If it is not recommended, then perhaps it should not be in Devdocs. Is there a good reason?

@dobooth  We know It's not a recommended way of overriding methods. But when we are doing customization on Magento for many projects, we end up like there is no other way than adding **preference**  for the particular class to achieve our requirements. 
That's why I would like to have an example of this topic. I believe this example will help the developer's on how to use preference for method overriding.

I appreciate your thoughts on this. @dobooth 

CC: @rogyar

**Comment by rogyar** [comment]
Hi @dobooth. Thank you for asking.

This part of extensibility is like a "last chance" to extend some functionality that works when no other recommended extensibility method applies. Overriding using preferences is a part of Magento extensibility system, that's why I think it's a good idea to document it but leave the note that it's not recommended approach until you have nothing else to do for the extending the functionality.

**Comment by dobooth** [comment]
running tests

**Pull request merged**